### PR TITLE
feat: added get_contents_channel (wl-paste --watch)

### LINKED ISSUE #5: Implement --watch
Feature parity with wl-clipboard 2.0.

### PR DESCRIPTION
**Pull request by H3rmt**
fixes https://github.com/YaLTeR/wl-clipboard-rs/issues/5

### CODE REVIEW

File: src/paste.rs [hunk 1]
```diff
@@ -384,3 +362,147 @@ pub(crate) fn get_contents_internal(
...
+) -> Result<mpsc::Receiver<Result<(PipeReader, String), Error>>, Error> {
+    get_contents_channel_internal(seat, mime_type, None)
+}
+
+pub(crate) fn get_contents_channel_internal(
+    seat: Seat<'static>,
+    mime_type: MimeType<'static>,
+    socket_name: Option<OsString>,
+) -> Result<mpsc::Receiver<Result<(PipeReader, String), Error>>, Error> {
+    let (rs, tx) = mpsc::channel();
```
Comment (YaLTeR) [inline src/paste.rs]: It's usually (tx, rx) (for "transfer" and "receive").

File: src/paste.rs [hunk 2]
```diff
@@ -384,3 +362,147 @@ pub(crate) fn get_contents_internal(
...
+    let (mut queue, mut state, _) = get_offer(false, seat, socket_name)?;
+    // drop first selections (only trigger on new)
+    state.offers.clear();
+    thread::spawn(move || {
+        loop {
+            match queue.blocking_dispatch(&mut state) {
+                Err(err) => {
+                    let _ = rs.send(Err(Error::WaylandCommunication(err)));
+                }
+                Ok(_) => match get_seat(&mut state, seat) {
```
Comment (YaLTeR) [inline src/paste.rs]: I don't think this should re-figure out the seat each iteration (and complain about NoSeats each iteration).

File: src/paste.rs [hunk 3]
```diff
@@ -384,3 +362,147 @@ pub(crate) fn get_contents_internal(
...
+) -> Result<mpsc::Receiver<Result<(PipeReader, String), Error>>, Error> {
+    let (rs, tx) = mpsc::channel();
+    let (mut queue, mut state, _) = get_offer(false, seat, socket_name)?;
+    // drop first selections (only trigger on new)
+    state.offers.clear();
+    thread::spawn(move || {
+        loop {
+            match queue.blocking_dispatch(&mut state) {
+                Err(err) => {
+                    let _ = rs.send(Err(Error::WaylandCommunication(err)));
```
Comment (YaLTeR) [inline src/paste.rs]: Shouldn't these errors break out of the loop?

File: src/paste.rs [hunk 4]
```diff
@@ -384,3 +362,147 @@ pub(crate) fn get_contents_internal(
...
+    get_contents_channel_internal(seat, mime_type, None)
+}
+
+pub(crate) fn get_contents_channel_internal(
+    seat: Seat<'static>,
+    mime_type: MimeType<'static>,
+    socket_name: Option<OsString>,
+) -> Result<mpsc::Receiver<Result<(PipeReader, String), Error>>, Error> {
+    let (rs, tx) = mpsc::channel();
+    let (mut queue, mut state, _) = get_offer(false, seat, socket_name)?;
```
Comment (YaLTeR) [inline src/paste.rs]: Maybe move this into the loop to avoid the outer `Result<>` entirely?

**Review by YaLTeR — CHANGES_REQUESTED** [review]
I would like to see tests in `src/tests/` covering this extensively with various edge cases.

I think I'd also like an implementation of `wl-paste --watch`. It will prove that the design works (or show that it doesn't work in some cases, which then should be fixed).

File: src/paste.rs (hunk 2 above)
Comment (H3rmt) [reply src/paste.rs]: It is not possible to hold a reference to the seat across the loop as the state is borrowed in blocking_dispatch

File: src/paste.rs (hunk 3 above)
Comment (H3rmt) [reply src/paste.rs]: I think it would be usefull to send these errors through the channel to react to errors. 

blocking_dispatch could in theory fail once and then work again (I think?)

The loop is now exited if any Err or Ok couldnt be send as the channel was closed

File: src/paste.rs (hunk 4 above)
Comment (H3rmt) [reply src/paste.rs]: I cant move the get_offer into the loop as I need the queue.

**Comment by H3rmt** [comment]
I need some time to add the tests but I fixed most of the changes you requested

File: src/paste.rs (hunk 4 above)
Comment (H3rmt) [reply src/paste.rs]: I could send these errors through the channel, but I want errors receved in the channel to be errors that dont stop the loop. 
If outer Result the recever will never get any data, while Results in the channel indicate reading the current value from the clipboard failed and future clipboard changes will still be read and send through the channel

File: src/tests/paste.rs [hunk 5]
```diff
@@ -310,6 +309,177 @@ fn get_contents_wrong_mime_type() {
...
+    let tx =
+        get_contents_channel_internal(Seat::Unspecified, MimeType::Any, Some(socket_name.clone()))
+            .expect("unable to create channel");
+
+    let mut result = tx.recv().expect("failed to receive").expect("no data");
+    assert_eq!(result.1, "application/octet-stream");
+
+    let mut contents = vec![];
+    result.0.read_to_end(&mut contents).unwrap();
+    assert_eq!(contents, [1, 3, 3, 7, 8]);
```
Comment (YaLTeR) [inline src/tests/paste.rs:358]: These tests are a good start. Additionally, I'd like to see testing copying and receiving multiple times (that's the whole point after all), also recovering after an intermediate error.

**Review by YaLTeR — COMMENTED** [review]
An impl of `wl-paste --watch` would be good too.

File: src/tests/paste.rs (hunk 5 above)
Comment (H3rmt) [reply src/tests/paste.rs:358]: I cant get tests to work with multiple copies, idk why it doesnt work

**Comment by eventualbuddha** [comment]
@H3rmt is this something you’re planning to continue working on? I’m interested in seeing `--watch` be added and would be willing to contribute some time to this.

**Comment by H3rmt** [comment]
I continued development a bit here: https://github.com/H3rmt/hyprshell/tree/hyprshell-release/dep-crates/wl-clipboard-rs

I just didnt have the time to upstream these changes. You could idealy contribute the --watch flag for the binary, as I am only interesed in using wl-clipboard as a library. 
The other blocker were the tests, I just couldnt get some to work. 
I will upstream the remaining changes I have done so you can start working on the integration with the binary

**Comment by eventualbuddha** [comment]
> I will upstream the remaining changes I have done so you can start working on the integration with the binary

Sounds great, thanks.

**Comment by H3rmt** [comment]
@eventualbuddha I published all the changes from my local fork to this branch.

**Comment by H3rmt** [comment]
superseded by https://github.com/YaLTeR/wl-clipboard-rs/pull/83
